Document behavior of JobConnection.{run_and_measure|expectation} for …

ntezak opened the pull request:
…non-deterministic programs

This follows our recent internal Slack discussion. 
The rendered docs look like this: 
![image](https://user-images.githubusercontent.com/1573961/32509024-055fb8f8-c3a1-11e7-83ec-b60b0cfa0d4a.png)

ntezak commented:
I hope the italicized parts are not too dramatic :)

tarballs-are-good commented:
These comments are good.

Can we go one step further and add to the documentation a section on "Optimized Calls"? The `run` method is the only true bonafide way to evaluate things. _Every single other_ way to evaluate a program has a defect when the program has side-effects, non-determinism, _including_ the `.wavefunction` call.

We can discuss that these are for optimization only, and have one of those little neat warning boxes.

willzeng commented:
@ntezak Text looks good. Like the italics. I agree with @tarballs-are-good that emphasizing that the non `run` calls are more for optimization and debugging on the qvm will be useful.  With this added we can merge.

ntezak commented:
Ok, I added a short subsection to the `JobConnection` section in `Getting Started`:
![image](https://user-images.githubusercontent.com/1573961/32531704-38e36662-c3fa-11e7-9e29-f7e122ebc654.png)

ntezak commented:
Ok, @willzeng 

- moved Optimized Calls up one toc-level, 
- converted method and class names to proper internal sphinx references to enable users to jump to their documentation
- Added note to the `expectation` api call
- Fixed another random RST formatting problem in `getting_started.rst`

Looks like this now:
![image](https://user-images.githubusercontent.com/1573961/32567121-28a072de-c46f-11e7-9047-41badf01a495.png)

ntezak commented:
Ok, @willzeng 

- moved Optimized Calls up one toc-level, 
- converted method and class names to proper internal sphinx references to enable users to jump to their documentation
- Added note to the `expectation` api call
- Fixed another random RST formatting problem in `getting_started.rst`

Looks like this now:
![image](https://user-images.githubusercontent.com/1573961/32567121-28a072de-c46f-11e7-9047-41badf01a495.png)

ntezak commented:
Ok, @willzeng 

- moved Optimized Calls up one toc-level, 
- converted method and class names to proper internal sphinx references to enable users to jump to their documentation
- Added note to the `expectation` api call
- Fixed another random RST formatting problem in `getting_started.rst`

Looks like this now:
![image](https://user-images.githubusercontent.com/1573961/32567121-28a072de-c46f-11e7-9047-41badf01a495.png)
